spring-integration-core/src/test/java/org/springframework/integration/support/leader/LockRegistryLeaderInitiatorTests.java

```diff
@@ -100,6 +107,58 @@ public void competing() throws Exception {
 		assertThat(another.getContext().isLeader(), is(true));
 	}
 
+	@Test
+	public void competingWithLock() throws Exception {
+		// set up first registry instance - this one will be able to obtain lock initially
+		LockRegistry firstRegistry = mock(LockRegistry.class);
+		Lock firstLock = mock(Lock.class);
+		given(firstRegistry.obtain(anyString())).willReturn(firstLock);
+		given(firstLock.tryLock(anyLong(), any(TimeUnit.class))).willReturn(true);
+
+		// set up first initiator instance using first LockRegistry
+		LockRegistryLeaderInitiator first =
+				new LockRegistryLeaderInitiator(firstRegistry, new DefaultCandidate());
+		CountDownLatch firstGranted = new CountDownLatch(1);
+		CountDownLatch firstRevoked = new CountDownLatch(1);
+		first.setLeaderEventPublisher(new CountingPublisher(firstGranted, firstRevoked));
+
+		// set up first registry instance - this one will NOT be able to obtain lock initially
+		LockRegistry secondRegistry = mock(LockRegistry.class);
+		Lock secondLock = mock(Lock.class);
+		given(secondRegistry.obtain(anyString())).willReturn(secondLock);
+		given(secondLock.tryLock(anyLong(), any(TimeUnit.class))).willReturn(false);
+
+		// set up second initiator instance using second LockRegistry
+		LockRegistryLeaderInitiator second =
+				new LockRegistryLeaderInitiator(secondRegistry, new DefaultCandidate());
+		CountDownLatch secondGranted = new CountDownLatch(1);
+		CountDownLatch secondRevoked = new CountDownLatch(1);
+		second.setLeaderEventPublisher(new CountingPublisher(secondGranted, secondRevoked));
+
+		// start initiators
+		first.start();
+		second.start();
+
+		Thread.sleep(1000);
+
+		// first initiator should lead and publish granted event
+		assertThat(first.getContext().isLeader(), is(true));
+		assertThat(second.getContext().isLeader(), is(false));
+		assertThat(firstGranted.await(10, TimeUnit.SECONDS), is(true));
+
+		// simulate first registry instance unable to obtain lock, for example due to lock timeout
+		given(firstLock.tryLock(anyLong(), any(TimeUnit.class))).willReturn(false);
```

Comment:
This is the problem with Mockito.
You can't re-stub the same method on mock.

I guess `Mockito.reset(firstLock)` might help us, but it isn't an advice from the Mockito itself. See `reset()` JavaDocs.

I think you should consider to implement an `Answer` for the first `given()` which will return `false` according some external condition, e.g. some `AtomicBoolean` variable change here. For example:
````java
AtomicBoolean firstLocked = new AtomicBoolean(true);
...
given(firstLock.tryLock(anyLong(), any(TimeUnit.class))).willAnswer(i -> firstLocked.get());
...
given(secondLock.tryLock(anyLong(), any(TimeUnit.class))).willAnswer(i -> !firstLocked.get());
...
// simulate first registry instance unable to obtain lock, for example due to lock timeout
firstLocked.set(false);
````

I also do this in your test:
```
first.setBusyWaitMillis(1);
first.setHeartBeatMillis(10);
...
Thread.sleep(100);
```
For better performance.